eureka-client/src/main/java/com/netflix/discovery/shared/transport/decorator/RetryableEurekaHttpClient.java

```diff
@@ -152,10 +159,14 @@ public void shutdown() {
         List<EurekaEndpoint> candidateHosts = clusterResolver.getClusterEndpoints();
         quarantineSet.retainAll(candidateHosts);
 
-        // If all hosts are bad, we have no choice but start over again
-        if (quarantineSet.size() == candidateHosts.size()) {
+        // If enough hosts are bad, we have no choice but start over again
+        int threshold = candidateHosts.size() * transportConfig.getRetryableClientQuarantineRefreshPercentage() / 100;
+        if (quarantineSet.isEmpty()) {
+            // no-op
+        } else if (quarantineSet.size() >= threshold) {
```

Comment:
Good point. However we should also also not have it so that if there is only a single good server left, then we send a thundering herd there.